elasticdl/pkg/ps/server.go

```diff
@@ -78,80 +78,117 @@ func (s *Server) reportModelVersionIfNeeded(modelVersion int32) {
+	s.versionLock.Lock()
+	s.Model.Version += int32(1)
+	s.versionLock.Unlock()
+	s.reportModelVersionIfNeeded(s.Model.Version)
+	var resp = proto.PushGradientsResponse{
+		Accepted: true,
+		Version:  s.Model.Version,
+	}
+	return &resp, err
 }
 
-// PushModel pushes model to server
-func (s *Server) PushModel(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
+// PushModel push Model to server
+func (s *Server) PushModel(ctx context.Context, in *proto.Model) (*empty.Empty, error) {
 	s.lock.Lock()
 	var err error
-	if !s.Param.InitStatus {
-		err = s.Param.InitFromModelPB(in)
+	if !s.Model.Initialized {
+		err = s.Model.InitFromModelPB(in)
 		if err == nil {
-			s.Param.InitStatus = true
+			s.Model.Initialized = true
 		}
 	}
 	s.lock.Unlock()
 	return &empty.Empty{}, err
 }
 
-// PushEmbeddingInfo pushes embedding info to server
-func (s *Server) PushEmbeddingInfo(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
-	s.lock.Lock()
-	err := s.Param.InitFromModelPB(in)
-	s.lock.Unlock()
-	return &empty.Empty{}, err
+/* DEPRECATED
+// PullVariable pulls variable from server
+func (s *Server) PullVariable(ctx context.Context, in *proto.PullVariableRequest) (*proto.PullVariableResponse, error) {
+	// TODO(qijun) only support async now
+	var res proto.PullVariableResponse
+	if !s.Model.Initialized {
+		res.ModelInitStatus = false
+		return &res, nil
+	}
+	res.ModelInitStatus = true
+	res.Model = &proto.Model{}
+	res.Model.Version = s.Model.Version
+	if s.Model.Version > in.CurrentModelVersion {
+		for name, tensor := range s.Model.DenseParameters {
+			res.Model.Param = append(res.Model.Param, &proto.Tensor{
+				Name: name,
+				Dim: tensor.Dims,
+				Content: tensor.Buffer,
+			})
+		}
+	}
+	s.Model.Initialized = true
+	return &res, nil
 }
 
-// PushGradients pushes gradient to server
-func (s *Server) PushGradients(ctx context.Context, in *pb.Model) (*pb.PushGradientsResponse, error) {
-	// TODO(qijun) only support async now
-	var res pb.PushGradientsResponse
-	var grads []*common.Tensor
-	for _, gradPB := range in.Param {
-		grad := common.DeserializeTensorPB(gradPB)
-		grads = append(grads, grad)
+// PullEmbeddingVector pulls embedding vector from server
+func (s *Server) PullEmbeddingVector(ctx context.Context, in *proto.PullEmbeddingVectorRequest) (*proto.Tensor, error) {
+	if in.Ids == nil {
+		return &proto.Tensor{}, nil
 	}
-	err := s.Opt.ApplyGradients(grads, s.Param)
-	s.versionLock.Lock()
-	s.Param.Version += int32(1)
-	s.versionLock.Unlock()
-	s.reportModelVersionIfNeeded(s.Param.Version)
-	res.Accepted = true
-	res.Version = s.Param.Version
-	return &res, err
+	table := s.Model.GetEmbeddingTable(in.Name)
+	if table == nil {
+		return &proto.Tensor{}, fmt.Errorf("Request embedding Table %s not found in Param", in.Name)
+	}
+	t := table.GetEmbeddingVectors(in.Ids)
+	return &proto.Tensor{
+		Name: in.Name,
+		Dim: t.Dims,
+		Content: t.Buffer,
+	}, nil
 }
 
+// PushEmbeddingInfo pushes embedding info to server
```

Comment:
Since worker is still using this RPC service, we have to preserve it in this PR.
We could clean it in next PR.

Reply:
fixed